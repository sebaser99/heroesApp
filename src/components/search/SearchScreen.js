import { useLocation, useNavigate } from "react-router-dom";
import { useForm } from "../../hooks/useForm/useForm"
import { getHeroesByName } from "../../selectors/getHeroesByName";
import { HeroCard } from "../hero/HeroCard";
import queryString from 'query-string'
import { useMemo } from "react";

export const SearchScreen = () => {
    const navigate = useNavigate();
    const location = useLocation()
    const {q = ''} = queryString.parse(location.search)

    const [formValue, handleChangeInput ] = useForm({
        searchText: q
    }) 
    
    const {searchText} = formValue;

    const handleSubmit = (e)=>{
        e.preventDefault()
        navigate(`?q=${searchText}`)
    }

    const filteredHeroes = useMemo(()=> getHeroesByName(q), [q]) 

    return (
        <div>
             <h1>Búsqueda</h1>
             <hr/>

             <div className="row">
                <div className="col-5">
                
                    <form onSubmit={handleSubmit}>
                        <input 
                            type="text"
                            placeholder="Busca tu heroe"
                            className="form-control"
                            name="searchText"
                            autoComplete="off"
                            value={searchText}
                            onChange={handleChangeInput}
                        />
                        <button className="btn btn-outline-dark mt-2 btn-block" 
                                type='submit'>
                            Buscar ...
                        </button>
                    </form>
                </div>

                <div className="col-7" >

                    {
                        (q === '' ) 
                        ? <div className="alert alert-warning">Realiza una búsqueda</div>
                        : (filteredHeroes.length === 0 ) && <div className="alert alert-danger">No hay resultados {q}</div>
                    }
                   
                    {
                        filteredHeroes.map(hero =>(
                            <HeroCard
                                key={hero.id}
                                {...hero}
                            />
                        ))
                    }
                </div>

             </div>
        </div>
    )
}
