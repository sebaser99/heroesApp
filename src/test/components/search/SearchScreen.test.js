import { mount } from "enzyme"
import { MemoryRouter } from "react-router-dom"
import { SearchScreen } from "../../../components/search/SearchScreen"

const mockNavigate = jest.fn();

jest.mock('react-router-dom', ()=>({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockNavigate
})) 
//este mock lo hago 

describe('vamos a probar <SearchScreen />', ()=>{

    test('debe hacer match con el snapshot', ()=>{
        const wrapper = mount(
        <MemoryRouter initialEntries={['/search']}>
            <SearchScreen />
        </MemoryRouter>)

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('.alert-info').text().trim()).toBe('Realiza una bÃºsqueda')
    })

    test('debe mostrar a batman y el input con el valor del query string', ()=>{
        const wrapper = mount(
        <MemoryRouter initialEntries={['/search?q=batman']}>
            <SearchScreen />
        </MemoryRouter>)

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('input').prop('value')).toBe('batman')
    })

    test('debe mostrar a un mensaje de error al enviar un heroe que no existe', ()=>{
        const wrapper = mount(
        <MemoryRouter initialEntries={['/search?q=batman444']}>
            <SearchScreen />
        </MemoryRouter>)

        
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('.alert-danger').text().trim()).toBe(`No hay resultados ${wrapper.find('input').prop('value')}`)
    })

    test('debe mostrar el llamado de la funcion handleSubmit', ()=>{
        const wrapper = mount(
        <MemoryRouter initialEntries={['/search']}>
            <SearchScreen />
        </MemoryRouter>)

        wrapper.find('input').simulate('change', {
           target: {
               name: 'searchText'   ,
               value: 'batman'
           }
        })
        wrapper.find('form').prop('onSubmit')({
            preventDefault(){}
        })

        expect(mockNavigate).toHaveBeenCalled()
        expect(mockNavigate).toHaveBeenCalledWith('?q=batman')
    })
})