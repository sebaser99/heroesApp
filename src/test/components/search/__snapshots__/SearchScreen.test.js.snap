// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`vamos a probar <SearchScreen /> debe hacer match con el snapshot 1`] = `
<div>
  <h1>
    Búsqueda
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Buscar
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Buscar"
          type="text"
          value=""
        />
        <button
          className="btn btn-outline-primary mt-1 btn-block"
          type="submit"
        >
          Buscar ...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <div
        className="alert alert-info"
      >
        Realiza una búsqueda
      </div>
    </div>
  </div>
</div>
`;

exports[`vamos a probar <SearchScreen /> debe mostrar a batman y el input con el valor del query string 1`] = `
<div>
  <h1>
    Búsqueda
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Buscar
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Buscar"
          type="text"
          value="batman"
        />
        <button
          className="btn btn-outline-primary mt-1 btn-block"
          type="submit"
        >
          Buscar ...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <div
        className="col"
      >
        <div
          className="card"
          style={
            Object {
              "height": 300,
            }
          }
        >
          <div
            className="row no-gutters"
          >
            <div
              className="col-4 "
            >
              <img
                alt="Batman"
                className="card-img-top"
                src="/statics/images/dc-batman.jpg"
              />
            </div>
            <div
              className="col-8"
            >
              <div
                className="card-body"
              >
                <h5
                  className="card-title"
                >
                  Batman
                </h5>
                <p
                  className="card-text"
                >
                  Bruce Wayne
                </p>
                <p
                  className="card-text"
                >
                  <small
                    className="text-muted"
                  >
                    Detective Comics #27
                  </small>
                </p>
                <a
                  href="/hero/dc-batman"
                  onClick={[Function]}
                >
                  Ver más ...
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`vamos a probar <SearchScreen /> debe mostrar a un mensaje de error al enviar un heroe que no existe 1`] = `
<div>
  <h1>
    Búsqueda
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Buscar
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Buscar"
          type="text"
          value="batman444"
        />
        <button
          className="btn btn-outline-primary mt-1 btn-block"
          type="submit"
        >
          Buscar ...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <div
        className="alert alert-danger"
      >
        No hay resultados 
        batman444
      </div>
    </div>
  </div>
</div>
`;
